File: src/main/java/tmi/ui/RecipeNode.java
package tmi.ui;

import arc.Core;
import arc.func.Cons2;
import arc.graphics.Color;
import arc.graphics.g2d.Draw;
import arc.graphics.g2d.Lines;
import arc.math.Mathf;
import arc.scene.Element;
import arc.scene.event.Touchable;
import arc.scene.ui.Button;
import arc.scene.ui.Tooltip;
import arc.scene.ui.layout.Scl;
import arc.scene.ui.layout.Table;
import arc.util.Scaling;
import arc.util.Time;
import mindustry.gen.Icon;
import mindustry.gen.Tex;
import mindustry.ui.Styles;
import tmi.recipe.RecipeItemStack;
import tmi.recipe.types.RecipeItem;

import static mindustry.Vars.mobile;
import static tmi.TooManyItems.binds;

/**在{@link tmi.recipe.RecipeType}进行布局时所操作的元素对象，用于显示单个条目信息和提供控制逻辑*/
public class RecipeNode extends Button {
  public static final float SIZE = Scl.scl(80);

  public final RecipeItemStack stack;

  public final NodeType type;

  float progress, alpha;
  boolean activity, touched;
  float time;
  int clicked;

  Cons2<RecipeItem<?>, RecipesDialog.Mode> click;

  public RecipeNode(NodeType type, RecipeItemStack stack, Cons2<RecipeItem<?>, RecipesDialog.Mode> click){
    this.type = type;
    this.click = click;

    setBackground(Tex.button);
    this.stack = stack;

    touchable = Touchable.enabled;

    defaults().padLeft(8).padRight(8);

    setSize(SIZE);

    addListener(new Tooltip(t -> t.add(stack.item().localizedName(), Styles.outlineLabel)){{ allowMobile = true; }});

    hovered(() -> activity = true);
    exited(() -> activity = false);
    tapped(() -> {
      touched = true;
      time = Time.globalTime;
    });
    released(() -> {
      touched = false;

      if (Time.globalTime - time < 12){
        if (!mobile || Core.settings.getBool("keyboard")) {
          click.get(stack.item(), Core.input.keyDown(binds.hotKey) ? type == NodeType.block ? RecipesDialog.Mode.factory : RecipesDialog.Mode.usage : RecipesDialog.Mode.recipe);
        }
        else {
          clicked++;
          if (clicked >= 2){
            click.get(stack.item(), type == NodeType.block ? RecipesDialog.Mode.factory : RecipesDialog.Mode.usage);
            clicked = 0;
          }
        }
      }
      else {
        if (stack.item.hasDetails() && progress >= 0.95f){
          stack.item.displayDetails();
        }
      }
    });

    stack(
        new Table(t -> t.image(stack.item.icon()).size(SIZE/Scl.scl()*0.62f).scaling(Scaling.fit)),

        new Table(t -> {
          t.left().bottom();
          t.add(stack.getAmount(), Styles.outlineLabel);
          t.pack();
        }),

        new Table(t -> {
          if (!stack.item.locked()) return;
          t.right().bottom().defaults().right().bottom().pad(4);
          t.image(Icon.lock).scaling(Scaling.fit).size(10).color(Color.lightGray);
        })
    ).grow().pad(5);
  }

  @Override
  public void act(float delta) {
    super.act(delta);

    alpha = Mathf.lerpDelta(alpha, touched || activity ? 1 : 0, 0.08f);
    progress = Mathf.approachDelta(progress, stack.item.hasDetails() && touched? 1.01f : 0, 1/60f);
    if (Time.globalTime - time > 12 && clicked == 1){
      click.get(stack.item(), RecipesDialog.Mode.recipe);
      clicked = 0;
    }
  }

  @Override
  protected void drawBackground(float x, float y) {
    super.drawBackground(x, y);
    Lines.stroke(Scl.scl(34), Color.lightGray);
    Draw.alpha(0.5f);

    Lines.arc(x + width/2, y + height/2, Scl.scl(18), progress, 90);
  }
}

File: src/main/kotlin/tmi/ui/RecipeNode.kt
package tmi.ui

import arc.Core
import arc.func.Cons
import arc.func.Cons3
import arc.graphics.Color
import arc.graphics.g2d.Draw
import arc.input.KeyCode
import arc.math.Mathf
import arc.scene.Element
import arc.scene.event.InputEvent
import arc.scene.event.InputListener
import arc.scene.event.Touchable
import arc.scene.ui.Button
import arc.scene.ui.Tooltip
import arc.scene.ui.layout.Scl
import arc.scene.ui.layout.Table
import arc.util.Scaling
import arc.util.Time
import mindustry.Vars
import mindustry.gen.Icon
import mindustry.gen.Sounds.click
import mindustry.gen.Tex
import mindustry.ui.Styles
import tmi.TooManyItems
import tmi.recipe.RecipeItemStack
import tmi.util.Shapes

val NODE_SIZE: Float = Scl.scl(80f)

/**在[tmi.recipe.RecipeType]进行布局时所操作的元素对象，用于显示单个条目信息和提供控制逻辑 */
class RecipeNode(
  val type: NodeType,
  val stack: RecipeItemStack<*>,
  val clickListener: Cons3<RecipeItemStack<*>, NodeType, RecipesDialog.Mode>? = null
) : Button() {
  private var lastTouchedTime = 0f
  private var progress: Float = 0f
  private var alpha: Float = 0f
  private var clicked = 0
  var activity: Boolean = false
  var touched: Boolean = false

  init {
    background = Tex.button
    touchable = Touchable.enabled

    defaults().padLeft(8f).padRight(8f)

    setSize(NODE_SIZE)

    addListener(object : Tooltip(Cons { t: Table -> t.add(stack.item.localizedName, Styles.outlineLabel) }) {
      init {
        allowMobile = true
      }
    })

    addListener(object : InputListener() {
      override fun enter(event: InputEvent?, x: Float, y: Float, pointer: Int, fromActor: Element?) {
        super.enter(event, x, y, pointer, fromActor)
        activity = true
      }

      override fun exit(event: InputEvent?, x: Float, y: Float, pointer: Int, toActor: Element?) {
        super.exit(event, x, y, pointer, toActor)
        activity = false
      }

      override fun touchDown(event: InputEvent?, x: Float, y: Float, pointer: Int, button: KeyCode?): Boolean {
        if (pointer != 0 && button != KeyCode.mouseLeft && button != KeyCode.mouseRight) return false
        lastTouchedTime = Time.globalTime
        touched = true
        activity = true
        return true
      }

      override fun touchUp(event: InputEvent?, x: Float, y: Float, pointer: Int, button: KeyCode?) {
        if (pointer != 0 && button != KeyCode.mouseLeft && button != KeyCode.mouseRight) return
        touched = false
        activity = false

        if (stack.item.hasDetails && (progress >= 0.95f || click == null)) {
          stack.item.displayDetails()
          progress = 0f

          return
        }

        if (clickListener != null && Time.globalTime - lastTouchedTime < 12) {
          if (!Vars.mobile || Core.settings.getBool("keyboard")) {
            clickListener.get(
              stack,
              type,
              if (button == KeyCode.mouseRight)
                if (type == NodeType.BLOCK) RecipesDialog.Mode.FACTORY
                else RecipesDialog.Mode.USAGE
              else RecipesDialog.Mode.RECIPE
            )
          }
          else {
            clicked++
            if (clicked >= 2) {
              clickListener.get(
                stack,
                type,
                if (type == NodeType.BLOCK) RecipesDialog.Mode.FACTORY
                else RecipesDialog.Mode.USAGE
              )
              clicked = 0
            }
          }
        }
      }
    })

    stack(
      Table { it.image(stack.item.icon).size(NODE_SIZE/Scl.scl()*0.62f).scaling(Scaling.fit) },

      Table {
        var last = false

        it.left().bottom()
        it.add(stack.getAmount(), Styles.outlineLabel)
          .apply {
            if (stack.alternativeFormat != null) update{ l ->
              val isDown = Core.input.keyDown(TooManyItems.binds.hotKey)
              if (last != isDown){
                l.setText(
                  if (isDown && stack.alternativeFormat != null) stack.alternativeFormat!!.format(stack.amount)
                  else stack.getAmount()
                )
                last = isDown
              }
            }
          }
        it.pack()
      },

      Table {
        if (!stack.item.locked) return@Table
        it.right().bottom().defaults().right().bottom().pad(4f)
        it.image(Icon.lock).scaling(Scaling.fit).size(10f).color(Color.lightGray)
      }
    ).grow().pad(5f)
  }

  override fun act(delta: Float) {
    super.act(delta)
    alpha = Mathf.lerpDelta(alpha, (if (touched || activity) 1 else 0).toFloat(), 0.08f)
    progress = Mathf.approachDelta(progress, if (stack.item.hasDetails && click != null && touched) 1f else 0f, 1/60f)

    if (clickListener != null && Time.globalTime - lastTouchedTime > 12 && clicked == 1) {
      clickListener.get(stack, type, RecipesDialog.Mode.RECIPE)
      clicked = 0
    }
  }

  override fun drawBackground(x: Float, y: Float) {
    super.drawBackground(x, y)
    Draw.color(Color.lightGray)
    Draw.alpha(0.5f)

    Shapes.fan(x + width/2, y + height/2, Scl.scl(32f), -progress*360, 90f)
  }
}
